File: src-tauri/src/commands/docx_edit.rs
use super::*;

/// Rewrite a Word file from edited text, paragraph by paragraph.
///
/// THE GAP THIS CLOSES: the AI could already edit a `.docx` in place —
/// `edit_file` has a run-split OOXML matcher that rewrites `word/document.xml`
/// while preserving every style, table and image. The USER could not. The Edit
/// button on a Word file offered "Edit as text", which edited the EXTRACTED
/// text and saved a separate `.md` copy, leaving the original untouched. The
/// person who owns the document had strictly less power over it than the model
/// working on their behalf.
///
/// The rewrite is per-paragraph on purpose. Replacing `word/document.xml`
/// wholesale from plain text would throw away everything that makes it a Word
/// file. Instead the paragraphs of the stored text are lined up against the
/// paragraphs of the edited text, and only the ones that actually differ are
/// pushed through the SAME matcher `edit_file` uses. An untouched paragraph is
/// not rewritten at all, so its runs, fonts and fields survive byte for byte.
///
/// Adding or removing paragraphs is refused rather than guessed at: with no
/// anchor to hang a new `<w:p>` on, "insert here" is not a question the text
/// alone can answer, and quietly dropping the change would be worse than
/// saying so.
#[tauri::command]
pub fn update_docx_text(
    app: tauri::AppHandle,
    state: State<'_, AppState>,
    id: String,
    content: String,
) -> Result<FileMeta, String> {
    let meta = state.with_room(|room| {
        let (name, _mime, bytes, extracted) = db::get_file_full(&room.conn, &id)?;
        let bytes = bytes.unwrap_or_default();
        if extraction::extension_of(&name) != "docx" {
            return Err(format!("\"{name}\" is not a Word document."));
        }
        let before = extracted.ok_or_else(|| {
            format!("\"{name}\" has no readable text, so there is nothing to edit.")
        })?;
        let patched = apply_paragraph_edits(&bytes, &before, &content)?;
        // The new searchable text is re-derived from the PATCHED FILE, never
        // taken from the editor's buffer: the two can legitimately differ (the
        // reader shows headers, footnotes and comments, which live in parts
        // this rewrite doesn't touch), and storing the buffer would quietly
        // delete them from the index.
        let text = extraction::extract_text(&name, &patched);
        store_file_bytes(&room.conn, &id, &patched, text.as_deref(), "You edited the document")?;
        db::get_file_meta(&room.conn, &id)
    })?;
    use tauri::Emitter;
    let _ = app.emit("room-files-changed", ());
    Ok(meta)
}

/// Paragraphs, as the reader sees them: non-empty lines, trimmed of the
/// trailing whitespace an editor adds.
fn paragraphs(text: &str) -> Vec<&str> {
    text.lines().map(str::trim_end).filter(|l| !l.trim().is_empty()).collect()
}

/// The core, split out so it is testable without a room.
pub(crate) fn apply_paragraph_edits(
    bytes: &[u8],
    before: &str,
    after: &str,
) -> Result<Vec<u8>, String> {
    let old_paras = paragraphs(before);
    let new_paras = paragraphs(after);
    if old_paras.len() != new_paras.len() {
        return Err(format!(
            "This editor can change the wording of a Word file's paragraphs, but not add or \
             remove them — the document has {} and the edited text has {}. Undo the added or \
             deleted line, or use \"Save as a copy\" to keep this version as a separate note.",
            old_paras.len(),
            new_paras.len()
        ));
    }
    let mut patched = bytes.to_vec();
    let mut changed = 0usize;
    for (old, new) in old_paras.iter().zip(new_paras.iter()) {
        if old == new {
            continue;
        }
        // The extracted text of a header, a footnote or a comment reaches the
        // editor too, but only `word/document.xml` is rewritten — so a
        // paragraph that lives elsewhere simply won't match. Say which one
        // rather than reporting a bare failure.
        let (next, count) = extraction::docx_replace_text(&patched, old, new).map_err(|e| {
            format!(
                "One paragraph could not be written back into the Word file, so nothing was \
                 saved. This happens when the changed line comes from a header, a footer, a \
                 footnote or a comment — those are shown in the reader but live outside the \
                 document body. Paragraph: \"{}\"\n\n{e}",
                snippet(old)
            )
        })?;
        // A paragraph repeated verbatim elsewhere in the document would be
        // rewritten in both places. Refuse instead of silently changing a line
        // the user never looked at.
        if count > 1 {
            return Err(format!(
                "\"{}\" appears {count} times in this document, so saving would change all of \
                 them. Edit it in a way that makes the line unique, or use \"Save as a copy\".",
                snippet(old)
            ));
        }
        patched = next;
        changed += 1;
    }
    if changed == 0 {
        return Err("Nothing changed.".into());
    }
    Ok(patched)
}

fn snippet(s: &str) -> String {
    let trimmed = s.trim();
    if trimmed.chars().count() <= 60 {
        return trimmed.to_string();
    }
    let cut: String = trimmed.chars().take(60).collect();
    format!("{cut}…")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::extraction::fake_office_zip;

    fn doc(paras: &[&str]) -> Vec<u8> {
        let body: String =
            paras.iter().map(|p| format!("<w:p><w:r><w:t>{p}</w:t></w:r></w:p>")).collect();
        fake_office_zip("word/document.xml", &format!("<w:document>{body}</w:document>"))
    }

    #[test]
    fn one_edited_paragraph_is_written_back_and_the_rest_untouched() {
        let bytes = doc(&["The fee is 5%.", "Payable monthly.", "Signed in Tel Aviv."]);
        let before = crate::extraction::extract_text("c.docx", &bytes).unwrap();
        let after = before.replace("The fee is 5%.", "The fee is 7%.");
        let patched = apply_paragraph_edits(&bytes, &before, &after).expect("edit refused");
        let text = crate::extraction::extract_text("c.docx", &patched).unwrap();
        assert!(text.contains("The fee is 7%."), "the edit did not land: {text}");
        assert!(text.contains("Payable monthly."), "an untouched paragraph was lost");
        assert!(text.contains("Signed in Tel Aviv."), "an untouched paragraph was lost");
        assert!(!text.contains("5%"), "the old wording survived: {text}");
    }

    #[test]
    fn several_paragraphs_can_change_in_one_save() {
        let bytes = doc(&["Alpha", "Beta", "Gamma"]);
        let before = crate::extraction::extract_text("c.docx", &bytes).unwrap();
        let after = before.replace("Alpha", "Alpha!").replace("Gamma", "Gamma!");
        let patched = apply_paragraph_edits(&bytes, &before, &after).unwrap();
        let text = crate::extraction::extract_text("c.docx", &patched).unwrap();
        assert!(text.contains("Alpha!") && text.contains("Gamma!"), "{text}");
        assert!(text.contains("Beta"), "the middle paragraph was disturbed: {text}");
    }

    #[test]
    fn adding_or_removing_a_paragraph_is_refused_in_plain_language() {
        let bytes = doc(&["One", "Two"]);
        let before = crate::extraction::extract_text("c.docx", &bytes).unwrap();
        let err = apply_paragraph_edits(&bytes, &before, &format!("{before}\nThree"))
            .expect_err("adding a paragraph should be refused");
        assert!(err.contains("not add or remove"), "unhelpful message: {err}");
        // …and the count is named, so the person can find what they added.
        assert!(err.contains('2') && err.contains('3'), "counts missing: {err}");
    }

    #[test]
    fn a_paragraph_repeated_verbatim_is_refused_rather_than_changed_twice() {
        // Editing one of two identical lines would silently rewrite BOTH — the
        // classic destructive find-and-replace.
        let bytes = doc(&["Confidential", "Body text", "Confidential"]);
        let before = crate::extraction::extract_text("c.docx", &bytes).unwrap();
        let after = before.replacen("Confidential", "Public", 1);
        let err = apply_paragraph_edits(&bytes, &before, &after)
            .expect_err("an ambiguous rewrite should be refused");
        assert!(err.contains("appears 2 times"), "unhelpful message: {err}");
        // Nothing was written: the caller only stores what this returns.
    }

    #[test]
    fn an_unchanged_save_is_not_treated_as_an_edit() {
        let bytes = doc(&["Same", "Same again"]);
        let before = crate::extraction::extract_text("c.docx", &bytes).unwrap();
        let err = apply_paragraph_edits(&bytes, &before, &before).expect_err("no-op accepted");
        assert_eq!(err, "Nothing changed.");
    }

    #[test]
    fn trailing_whitespace_and_blank_lines_are_not_edits() {
        // An editor adds a trailing newline on save; that must not be read as
        // "a paragraph was added".
        let bytes = doc(&["Alpha", "Beta"]);
        let before = crate::extraction::extract_text("c.docx", &bytes).unwrap();
        let after = format!("{}   \n\n", before.replace("Beta", "Beta!"));
        let patched = apply_paragraph_edits(&bytes, &before, &after).unwrap();
        let text = crate::extraction::extract_text("c.docx", &patched).unwrap();
        assert!(text.contains("Beta!"), "{text}");
    }
}

File: electron-migration/electron-app/electron/main/docxEdit.ts
/**
 * Port of `src-tauri/src/commands/docx_edit.rs` (231 lines, read in full) —
 * the single `update_docx_text` command behind the Word-file editor's Save
 * button.
 *
 * THE GAP THIS CLOSES (the Rust source's own framing, kept verbatim because
 * it is the reason this command exists at all): the AI could already edit a
 * `.docx` in place through `edit_file`'s run-split OOXML matcher, which
 * rewrites `word/document.xml` while preserving every style, table and
 * image. The USER could not — the Edit button on a Word file offered "Edit
 * as text", which edited the EXTRACTED text and saved a separate `.md` copy,
 * leaving the original untouched. This command gives the person who owns the
 * document the same power over it the model already had.
 *
 * PER-PARAGRAPH ON PURPOSE. Replacing `word/document.xml` wholesale from
 * plain text would throw away everything that makes it a Word file. Instead
 * the paragraphs of the stored text are lined up against the paragraphs of
 * the edited text, and only the ones that actually differ are pushed through
 * the SAME matcher the model's `edit_file` uses — {@link docxReplaceText}
 * (`editMatchDocx.ts`, already ported from `extraction/docx.rs`). An
 * untouched paragraph is not rewritten at all, so its runs, fonts and fields
 * survive byte for byte. Adding or removing a paragraph is refused rather
 * than guessed at: with no anchor to hang a new `<w:p>` on, "insert here" is
 * not a question the text alone can answer.
 *
 * REUSES RATHER THAN DUPLICATES, per this batch's instructions:
 *   - {@link docxReplaceText} (`editMatchDocx.ts`) — the docx run-split
 *     matcher itself. Nothing here parses OOXML or a zip directly.
 *   - {@link extractText}/{@link storeFileBytes} (`editMatch.ts`) — the same
 *     `extraction::extract_text` dispatcher and `commands::files::
 *     store_file_bytes` single-write-path `commit_plans` itself uses, so a
 *     manual Word edit re-derives its searchable text and snapshots its
 *     version history exactly the way an AI edit does.
 *   - {@link extensionOf} (`editMatchExtraction.ts`), {@link getFileFull}/
 *     {@link getFileMeta} (`db-host/files.ts`).
 *
 * NOT PORTED: `edit_gate.rs` (Wave 2's diff-preview approval gate) is a
 * SEPARATE Rust module this command never calls — confirmed by reading
 * `docx_edit.rs` in full; `update_docx_text` writes directly through
 * `store_file_bytes` under the room lock, with no `gated_write` wrapper.
 * `edit_gate.rs` exists in this codebase only for the tool-invoked mutations
 * `agent.rs`'s `exec_tool` dispatches (`edit_file`/`edit_files`/`write_file`/
 * `set_cells`); it has no bearing on this file.
 *
 * NOT A MODEL TOOL: `update_docx_text` has no `exec_tool` arm in `agent.rs`
 * and no entry in `toolSpecs.ts`/`toolSchema.ts` (confirmed by grep) — it
 * exists purely for the built-in Word-file viewer's Save button, exactly
 * like the Rust `#[tauri::command]` it ports. Nothing in this file touches
 * `execTool.ts`.
 *
 * THE UNCONDITIONAL EMIT, preserved exactly as the Rust source has it: the
 * "room-files-changed" broadcast fires AFTER the room-locked body returns —
 * even down the no-op branch, where the edited text said exactly what the
 * document already said and nothing was written. That is the Rust source's
 * own control flow (the emit sits outside the closure `state.with_room` runs,
 * not inside either of its two return points), not an oversight introduced by
 * this port.
 *
 * NO IPC WIRING in this batch, same as `dictStopTimeout.ts`/`recIpc.ts`/
 * `previewTools.ts`: Phase 2 (renderer/preload) needs an explicit owner
 * go-ahead before touching the live shipping app. {@link registerDocxEditIpc}
 * exists, ready to be wired, but nothing calls it yet. `update_docx_text` is
 * already declared on `ipc-contract.ts`'s wire contract (`src/api.ts:237`'s
 * `invoke<FileMeta>("update_docx_text", { id, content })`), so the channel
 * name and argument shape need no invention here.
 */

import type { IpcMain, IpcMainInvokeEvent } from "electron";
import type Database from "better-sqlite3-multiple-ciphers";
import { getFileFull, getFileMeta, type FileMeta } from "./db-host/files.js";
import { extensionOf, isUnicodeWhitespace } from "./editMatchExtraction.js";
import { extractText, storeFileBytes } from "./editMatch.js";
import { docxReplaceText } from "./editMatchDocx.js";
import type { OpenRoom } from "./turnEngine.js";

// ------------------------------------------------------------- room/emit plumbing

/** The slice of the (not-yet-ported) `AppState` this command needs: whichever
 * room is open RIGHT NOW. Reuses `turnEngine.ts`'s {@link OpenRoom} rather
 * than inventing a second "how do I reach the open room" convention — same
 * reasoning `recIpc.ts`'s/`previewTools.ts`'s own `RoomSource` gives. */
export interface RoomSource {
  currentRoom(): OpenRoom | null;
}

/** `AppState::with_room`'s own refusal, spelled the way `recIpc.ts` and
 * `previewTools.ts` already spell it. */
const NO_ROOM_OPEN = "No room is open.";

function openDb(room: RoomSource): Database.Database {
  const open = room.currentRoom();
  if (open === null) {
    throw new Error(NO_ROOM_OPEN);
  }
  return open.db;
}

/** `let _ = app.emit(...)` — a best-effort UI notification that must never
 * turn a successful save into a failed one. Same narrowest-possible contract
 * as `organizeTools.ts`'s/`safetyTools.ts`'s own `EmitFn`. */
export type EmitFn = (event: string, payload: unknown) => void;

function emitSafely(emit: EmitFn | undefined, event: string, payload: unknown): void {
  try {
    emit?.(event, payload);
  } catch {
    // Swallowed deliberately, matching Rust's `let _ = app.emit(...)`.
  }
}

// --------------------------------------------------------------------- paragraphs

/** Rust's `str::lines()`: split on `\n`, drop a trailing `\r` from each line,
 * and produce NO final empty line for a string that ends in `\n`. A local
 * copy of `editMatchExtraction.ts`'s own (unexported) `linesOf` — the same
 * duplication choice `organizeTools.ts`'s local `extensionOf` documents,
 * until a future batch collapses these onto one shared module. */
function linesOf(s: string): string[] {
  if (s === "") {
    return [];
  }
  const parts = s.split("\n");
  if (parts[parts.length - 1] === "") {
    parts.pop();
  }
  return parts.map((l) => (l.endsWith("\r") ? l.slice(0, -1) : l));
}

/** Rust's `str::trim_end()` — trims trailing `char::is_whitespace()`, the
 * Unicode `White_Space` property {@link isUnicodeWhitespace} already reads.
 * Deliberately NOT JS's native `.trimEnd()`, which both misses U+0085 (NEL)
 * and trims U+FEFF (a BOM Rust treats as an ordinary character) — the same
 * divergence `editMatchDocx.ts`'s `hasEdgeWhitespace` documents. Every
 * `White_Space` code point is in the BMP, so per-UTF-16-unit indexing is
 * exact here; no surrogate pair is ever a whitespace character. */
function trimEndUnicode(s: string): string {
  let end = s.length;
  while (end > 0 && isUnicodeWhitespace(s.charAt(end - 1))) {
    end -= 1;
  }
  return s.slice(0, end);
}

/** Rust's `str::trim()` — both ends, same Unicode property as
 * {@link trimEndUnicode}. */
function trimUnicode(s: string): string {
  let start = 0;
  let end = s.length;
  while (start < end && isUnicodeWhitespace(s.charAt(start))) {
    start += 1;
  }
  while (end > start && isUnicodeWhitespace(s.charAt(end - 1))) {
    end -= 1;
  }
  return s.slice(start, end);
}

/** Paragraphs, as the reader sees them: non-empty lines, trimmed of the
 * trailing whitespace an editor adds. Ported verbatim from
 * `docx_edit::paragraphs`. */
function paragraphs(text: string): string[] {
  return linesOf(text)
    .map(trimEndUnicode)
    .filter((l) => trimUnicode(l) !== "");
}

/** A short, character-count-safe (not UTF-16-unit-safe) preview of a
 * paragraph for an error message — long enough to identify the line, short
 * enough not to dump a whole document into a dialog. Ported verbatim from
 * `docx_edit::snippet`; `[...s]` gives code-point iteration the same way
 * Rust's `chars()` does, matching `editMatchDocx.ts`'s own convention for
 * character-indexed text. */
function snippet(s: string): string {
  const trimmed = trimUnicode(s);
  const chars = [...trimmed];
  if (chars.length <= 60) {
    return trimmed;
  }
  return `${chars.slice(0, 60).join("")}…`;
}

/**
 * The core, split out so it is testable without a room. `null` is "the
 * edited text says exactly what the document already says" — reporting that
 * as an error made a whitespace-only change impossible to close: Save
 * answered "Could not save — Nothing changed.", and the unsaved-edits dialog
 * reads a failed save as "your edit is still in the editor" and stays up,
 * leaving Discard as the only way out. Ported verbatim from
 * `docx_edit::apply_paragraph_edits`; throws (matching Rust's `Result<_,
 * String>` propagated through `?` at the command layer) rather than
 * returning a discriminated union, since — unlike `edit_match.rs`'s
 * tool-facing errors — nothing here needs a content-free `outcome` tag.
 */
export function applyParagraphEdits(bytes: Uint8Array, before: string, after: string): Buffer | null {
  const oldParas = paragraphs(before);
  const newParas = paragraphs(after);
  if (oldParas.length !== newParas.length) {
    throw new Error(
      "This editor can change the wording of a Word file's paragraphs, but not add or " +
        `remove them — the document has ${oldParas.length} and the edited text has ` +
        `${newParas.length}. Undo the added or deleted line, or use "Save as a copy" to keep ` +
        "this version as a separate note."
    );
  }
  let patched: Uint8Array = bytes;
  let changed = 0;
  for (let i = 0; i < oldParas.length; i++) {
    const oldPara = oldParas[i]!;
    const newPara = newParas[i]!;
    if (oldPara === newPara) {
      continue;
    }
    // The extracted text of a header, a footnote or a comment reaches the
    // editor too, but only `word/document.xml` is rewritten — so a paragraph
    // that lives elsewhere simply won't match. Say which one rather than
    // reporting a bare failure.
    const result = docxReplaceText(patched, oldPara, newPara);
    if (!result.ok) {
      throw new Error(
        "One paragraph could not be written back into the Word file, so nothing was " +
          "saved. This happens when the changed line comes from a header, a footer, a " +
          "footnote or a comment — those are shown in the reader but live outside the " +
          `document body. Paragraph: "${snippet(oldPara)}"\n\n${result.error}`
      );
    }
    // A paragraph repeated verbatim elsewhere in the document would be
    // rewritten in both places. Refuse instead of silently changing a line
    // the user never looked at.
    if (result.count > 1) {
      throw new Error(
        `"${snippet(oldPara)}" appears ${result.count} times in this document, so saving ` +
          'would change all of them. Edit it in a way that makes the line unique, or use ' +
          '"Save as a copy".'
      );
    }
    patched = result.bytes;
    changed += 1;
  }
  if (changed === 0) {
    return null;
  }
  return Buffer.isBuffer(patched) ? patched : Buffer.from(patched);
}

// ---------------------------------------------------------------------- command

/**
 * Rewrite a Word file from edited text, paragraph by paragraph. Ported from
 * `docx_edit::update_docx_text`, minus the Tauri wrapper — takes the room's
 * already-resolved `Database.Database`, matching `previewTools.ts`'s/
 * `recIpc.ts`'s convention.
 *
 * Throws (matching the Rust `?` on `db::get_file_full`) when `id` names no
 * file in this room; throws the same plain-language refusals the Rust source
 * does for a non-.docx file, a file with no readable text, an edit that adds
 * or removes a paragraph, an edit that lands in a header/footer/footnote/
 * comment, and an edit that would touch a repeated paragraph more than once.
 *
 * The "room-files-changed" broadcast is UNCONDITIONAL — it fires even when
 * nothing was written (the no-op branch), because that is what the Rust
 * source does: the emit sits after `state.with_room` returns, not inside
 * either of its two return points.
 */
export function updateDocxText(db: Database.Database, id: string, content: string, emit?: EmitFn): FileMeta {
  const [name, , bytesRaw, extracted] = getFileFull(db, id);
  const bytes = bytesRaw ?? Buffer.alloc(0);
  if (extensionOf(name) !== "docx") {
    throw new Error(`"${name}" is not a Word document.`);
  }
  if (extracted === null) {
    throw new Error(`"${name}" has no readable text, so there is nothing to edit.`);
  }
  const patched = applyParagraphEdits(bytes, extracted, content);
  let meta: FileMeta;
  if (patched === null) {
    // The buffer says what the document already says. Nothing to write, and
    // no version to record — but this is a save that succeeded, so the
    // editor's dirty flag clears and the close the user asked for goes
    // through.
    meta = getFileMeta(db, id);
  } else {
    // The new searchable text is re-derived from the PATCHED FILE, never
    // taken from the editor's buffer: the two can legitimately differ (the
    // reader shows headers, footnotes and comments, which live in parts this
    // rewrite doesn't touch), and storing the buffer would quietly delete
    // them from the index.
    const text = extractText(name, patched);
    storeFileBytes(db, id, patched, text, "You edited the document");
    meta = getFileMeta(db, id);
  }
  // Every OTHER way of changing a file broadcasts this; the plain Save in the
  // built-in editor did not, so the Library, the Scripts index and the front
  // page all kept showing what the file used to say.
  emitSafely(emit, "room-files-changed", undefined);
  return meta;
}

/**
 * Registers {@link updateDocxText} on the `update_docx_text` channel — the
 * Rust `#[tauri::command]` name `src/api.ts`'s `invoke("update_docx_text", …)`
 * already uses, so the renderer side needs no rename. `ipcMain` is accepted
 * as a parameter, typed against the real `electron` module without importing
 * it at runtime, so this file resolves and tests under plain Node/vitest
 * exactly like `registerPreviewIpc`/`registerRecIpc` do.
 *
 * Exported and directly testable, but — same as `previewTools.ts`/
 * `recIpc.ts` — NOT called from any live main-process entrypoint by this
 * batch. Wiring it in is Phase 2 work pending an explicit owner go-ahead.
 */
export function registerDocxEditIpc(
  ipcMain: Pick<IpcMain, "handle">,
  room: RoomSource,
  emit?: EmitFn
): void {
  ipcMain.handle(
    "update_docx_text",
    (_event: IpcMainInvokeEvent, args: { id: string; content: string }) =>
      updateDocxText(openDb(room), args.id, args.content, emit)
  );
}
